<!--<img src='./figures/logo-ecole-polytechnique-ve.jpg' style='position:absolute; top:0; right:0;' width='100px' height='' alt='' />-->

<center>**Bachelor of Ecole Polytechnique**</center>
<center>Computational Mathematics, year 2, semester 1</center>
<center>Lecturer: Lucas Gerin <a href="mailto:lucas.gerin@polytechnique.edu">(send mail)</a></center>

# Symbolic computation 3: Test


## Table of contents

- [Exercise 1: A polynomial riddle](#exo_web) 
- [Exercise 2: Rational fractions](#continued)
- [Exercise 3: Multivariate polynomial](#multivariate)
- [Exercise 4: Generating Functions](#GF)



In [1]:
# execute this part to modify the css style
from IPython.core.display import HTML
def css_styling():
    styles = open("./style/custom2.css").read()
    return HTML(styles)
css_styling()


In [2]:
## loading python libraries

# necessary to display plots inline:
%matplotlib inline   

# load the libraries
import matplotlib.pyplot as plt # 2D plotting library
import numpy as np              # package for scientific computing  
from pylab import *

from math import *              # package for mathematics (pi, arctan, sqrt, factorial ...)
import sympy as sympy             # package for symbolic computation
from sympy import *


# Please read!

<br><br>

<div markdown=1 class=Abstract>

### Guidelines for today

* You will use SymPy to solve the above exercises. The whole Notebook can be done with the following SymPy functions:
```
coeff(),expand(),factor(),Rational(),series(),simplify(),solve()
```
* In order to check your formulas do not hesitate to use `N()` to obtain a numerical evaluation.
* Remember that in SymPy you have to use `Rational` in order to prevent numerical evaluation of fractions. When you use symbolic variables, do not use `Rational`. For example: write `a/b` instead of `Rational(a,b)`.

## <a id='exo_web'></a>
### Exercise 1. A polynomial riddle

<div markdown=1 class="DoIt"> Let P be a polynomial of degree $3$ of the form $P(x)=x^3+ax^2+bx+c$.

Find the coefficient $a,b,c$ such that $P(1)=0$, $P(2)=3$ and $P(-1)=-5$.

SyntaxError: invalid syntax (676505452.py, line 3)

<div markdown=1 class="Answers"> 

$ a = \frac{-11}{6}, b = \frac{3}{2}, c = \frac{-2}{3} $

## <a id='continued'></a>
### Exercise 2. Rational fractions.

<div markdown=1 class="DoIt"> 
For $n\geq1$,
$$
    u_n=\sum_{k=1}^n\frac{1}{k(k+1)}.
$$

1. Using Sympy with Rational and a for loop, compute $u_n$ exactly for $n=5,10,20,30$. 
*(You should use the above expression without simplifying it and not use the tools from Question 3.)*

2. Based on these values, guess a closed form for $u_n$ as a rational function of $n$.
3. Using the sympy function summation (or Sum(...).doit()), compute $u_n$ in symbolic form and prove the conjecture made in Question 2. Find the description of the latter functions in [the sympy documentation](https://docs.sympy.org/latest/modules/concrete.html#sympy.concrete.summations.summation). 
</div>

In [12]:
#Question 1

def u (n):
    sum = Rational(0)
    for k in range (1, n+1):
        sum += Rational(1, k * (k+1))
    return sum

for n in range (5,35,5):
    print(u(n))

for n in range(20):
    print(u(n))




5/6
10/11
15/16
20/21
25/26
30/31
0
1/2
2/3
3/4
4/5
5/6
6/7
7/8
8/9
9/10
10/11
11/12
12/13
13/14
14/15
15/16
16/17
17/18
18/19
19/20


<div markdown=1 class="Answers to Question 2"> 

Closed form guess:
$ u_n = \frac{n}{n+1} $


In [70]:
#Question 3

def term (k):
    return Rational (1,k*(k+1))

from sympy import summation, oo, symbols, log

k, n, m = symbols('k n m', integer=True)

summation(1/((k+1)*k),(k,1,n)) # confirmed


1 - 1/(n + 1)

## <a id='multivariate'></a>
### Exercise 3. Multivariate Polynomials 

<div markdown=1 class="DoIt"> Consider the function
$$f(x,y)=\frac{e^{x+2y}}{1-x-y}.$$
Extract the coefficient of the monomial $x^2y^2$ from the Taylor expansion of $f$ around $(0,0)$.

In [26]:
var('x')
var('y')

f = (e ** (x + 2*y) / (1 - x - y))

display(f.series(x,0,3).coeff(x**2).series(y,0,3).coeff(y**2))

25.5000000000000

<div markdown=1 class="Answers"> 
the coeff of $x^2 * y^2$ is 25.5

## <a id="GF"></a>
### Exercise 4. Solving differential equation with Generating Functions

<div markdown=1 class="DoIt"> 
We consider the following linear differential equation:
$$
    y''(x)-xy'(x)-y(x)=0.
$$
We look for a solution in the form of a power series around $x=0$ with non-negative coefficients:
$$
    y(x)=\sum_{n=0}^{\infty}a_nx^n.
$$

1. Find the expressions of $y'$ and $y''$ as series. Use them and the differential equation to show that, for $n\geq2$,  $$a_{n+2}=\frac{a_n}{n+2}.$$ Justify you answer.

For Questions 2. and 3., we assume that $y(0)=1$ and $y'(0)=0$.

2. Find $a_0$ and $a_1$ such that the above conditions are satisfied. Compute the subsequent first terms $a_2,\dots,a_{12}$.

3. Find (without sympy) an analytical formula for $y$. Use sympy to compute its Taylor expansion and compare with the coefficients you computed in Question 2.

For Questions 4. and 5., $a_0$ and $a_1$ are unknown and will be represented using sympy variables.

4. Define a function, that takes an interger $N$ as an argument and outputs $P_N$ the truncated series defined by
$$P_N(x)=\sum_{n=0}^Na_nx^n +O(x^{N+1}).$$

5. For $N=5,10,100$, use the latter function and the sympy function diff (e.g., diff$(P,x)$ returns the derivative of $P$ with respect to $x$) to check that $P_N$ from the previous question satisfies the differential equation up to an error of order $O(x^{N-1})$.


<div markdown=1 class="Answers"> 

$ y' = \sum_{n=1}^{\infty} n \cdot a_n x^{n-1}$ 


$ y'' = \sum_{n=2}^{\infty} n (n-1) \cdot a_n x^{n-2} $

now, if we look at the coefficient of $x^n$ in the equation, we get
$x^n \cdot ( (n+1)n \cdot a_{n+2} - n  a_n - a_n) = 0$ 

=> $ a_{n+2} = a_n \cdot \frac{n+1}{(n+1)(n+2)} $ =>  $ a_{n+2} = \frac{a_n}{n+2} $

<div markdown=1 class="Answers."> 


In [31]:
#Question 2

a = [1,0]
for i in range(2,13):
    a.append(a[-2] / i)
    

print(a)


[1, 0, 0.5, 0.0, 0.125, 0.0, 0.020833333333333332, 0.0, 0.0026041666666666665, 0.0, 0.00026041666666666666, 0.0, 2.170138888888889e-05]


<div markdown=1 class="Answers"> 

$y(0) = 1$ and $y'(0) = 0$ => $a_0 = 1$ and  $a_1 = 0 $

Q3

$ y(x) = e^{x^2/2} $ => $ y'(x) = x * e^{x^2/2} $ => $ y''(x) = e^{x^2/2} + x^2*e^{x^2/2} $

In [33]:
#Question 3
var('x')
f = e ** (x**2/2)

f.series(x,0,12)


1 + 0.5*x**2 + 0.125*x**4 + 0.0208333333333333*x**6 + 0.00260416666666667*x**8 + 0.000260416666666667*x**10 + O(x**12)

In [68]:
#Question 4

var('a0')
var('a1')

def P_taylor_exp(N, fn=f):
    return fn.series(x,0,N+1)

def P(N):
    a = [a0,a1]
    ret = 0
    for i in range(2,N+1):
        a.append(a[-2] * Rational(1 , i))
    for i, a_i in enumerate(a):
        ret = ret + a_i * x ** i
    ret += O(x ** (N+1))
    return ret
    
#ret = sympy.sum(a_i * x ** i for i,a_i in enumerate(a)) + O(x ** (N+1))

display(P(15))

a1*x + a1*x**3/3 + a1*x**5/15 + a1*x**7/105 + a1*x**9/945 + a1*x**11/10395 + a1*x**13/135135 + a1*x**15/2027025 + a0 + a0*x**2/2 + a0*x**4/8 + a0*x**6/48 + a0*x**8/384 + a0*x**10/3840 + a0*x**12/46080 + a0*x**14/645120 + O(x**16)

In [69]:
#Question 5

def check_sat (P):
    return diff(diff(P,x),x) - x * diff(P,x) - P
    
N = [5,10,100]

for n in N:
    print( check_sat(P(n)).simplify() )


O(x**4)
O(x**9)
O(x**99)


In [65]:
#remark about question 5:

# my final change was to keep the coefficients as rational. without this, the taylor series would always generate some error 
#(for example ~1e-17 for x**6)